rotkehlchen/chain/ethereum/modules/cowswap/decoder.py

```diff
@@ -0,0 +1,236 @@
+
+GPV2_SETTLEMENT_ADDRESS = string_to_evm_address('0x9008D19f58AAbD9eD0D60971565AA8510560ab41')
+ETH_FLOW_ADDRESS = string_to_evm_address('0x40A50cf069e992AA4536211B23F286eF88752187')
+
+
+class TradeData(NamedTuple):
+    from_asset: 'CryptoAsset'
+    from_amount: FVal
+    to_asset: 'CryptoAsset'
+    to_amount: FVal
+
+
+class CowswapDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            ethereum_inquirer: 'EthereumInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=ethereum_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.eth = A_ETH.resolve_to_crypto_asset()
+
+    def _decode_eth_order_place_invalidate(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: list[ActionItem],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        if tx_log.topics[0] == PLACE_ETH_ORDER_SIGNATURE:
+            target_token_address = hex_or_bytes_to_address(tx_log.data[32:64])
+            target_token = EvmToken(evm_address_to_identifier(
+                address=target_token_address,
+                chain_id=ChainID.ETHEREUM,
+                token_type=EvmTokenKind.ERC20,
+            ))
+            for event in decoded_events:
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.asset == A_ETH and
+                    event.counterparty == ETH_FLOW_ADDRESS
+                ):
+                    event.event_type = HistoryEventType.DEPOSIT
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                    event.notes = f'Deposit {event.balance.amount} ETH to swap it for {target_token.symbol} in {CPT_COWSWAP}'  # noqa: E501
+                    event.counterparty = CPT_COWSWAP
+
+        elif tx_log.topics[0] in (INVALIDATE_ETH_ORDER_SIGNATURE, REFUND_ETH_ORDER_SIGNATURE):
+            for event in decoded_events:
+                if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.asset == A_ETH and
+                    event.counterparty == ETH_FLOW_ADDRESS
+                ):
+                    event.event_type = HistoryEventType.WITHDRAWAL
+                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                    event.counterparty = CPT_COWSWAP
+                    if tx_log.topics[0] == INVALIDATE_ETH_ORDER_SIGNATURE:
+                        event.notes = f'Invalidate an order that intended to swap {event.balance.amount} ETH in {CPT_COWSWAP}'  # noqa: E501
```

Comment:
Mmm well, technically yes, orders invalidations are initiated by third-parties, not by the users.

But since users deposit ETH to this contract, I thought that receiving ETH back would be better to be called a withdrawal.

Reply:
Also 

> Mmm well, technically yes, orders invalidations are initiated by third-parties, not by the users

Are you sure? At least from the on-chain example you have in the test the invalidator seems to be the same as the receiver: https://etherscan.io/tx/0x5769b4634ae26ec93aebc80a50e0676b0793af485041b249652bd7ee6703a9f5

Reply:
> Also
> 
> > Mmm well, technically yes, orders invalidations are initiated by third-parties, not by the users
> 
> Are you sure? At least from the on-chain example you have in the test the invalidator seems to be the same as the receiver: https://etherscan.io/tx/0x5769b4634ae26ec93aebc80a50e0676b0793af485041b249652bd7ee6703a9f5

https://docs.cow.fi/smart-contracts/eth-flow-contract/order-invalidation orders can be invalidated by both users and third parties if an order has expired. What I meant is that we can't assume that orders are always invalidated by the users.

Reply:
> Also
> 
> > Mmm well, technically yes, orders invalidations are initiated by third-parties, not by the users
> 
> Are you sure? At least from the on-chain example you have in the test the invalidator seems to be the same as the receiver: https://etherscan.io/tx/0x5769b4634ae26ec93aebc80a50e0676b0793af485041b249652bd7ee6703a9f5

Yeah, I did not say it correctly. The situation is the following:
1. If an order has not expired, it can be invalidated only by its creator https://docs.cow.fi/smart-contracts/eth-flow-contract/order-invalidation
2. If an order has not expired, in case of invalidation an `OrderInvalidation` event is emitted
3. If an order has expired, the invaliadtion can be initiated by any address (in fact it is almost always initiated by this address https://etherscan.io/address/0x0214ae5fd178986fa18ff792e0b995dc6a78cd56)
4. If an order has expired, an `OrderRefund` event is emitted https://github.com/cowprotocol/ethflowcontract/blob/main/src/CoWSwapEthFlow.sol#L190

So yeah, invalidations are initiated by the users and refunds are initiated by third-parties.